Add support for Python 3.12

daniel-zullo-frequenz opened the pull request:
The CI now runs the tests against Python 3.12 as well.

The Dockerfile for arm64 was added to the repository to run the tests for python 3.12 in arm64.

daniel-zullo-frequenz committed:
Add support for Python 3.12

The CI now runs the tests against Python 3.12 as well.

The Dockerfile for arm64 was added to the repository
to run the tests for python 3.12 in arm64.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Update dependencies for Python 3.12

Newer versions of grpcio, grpcio-tools and numpy are required
to run pytest_min tests with Python 3.12.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Update test installation for Python 3.12

The test installation now includes Python 3.12.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz commented:
Most of the changes in this draft need to be pushed in https://github.com/frequenz-floss/frequenz-repo-config-python instead, excepts for dependencies updates.

llucax reviewed (commented):
LGTM in general. I'm a bit concerned about having a CI that needs to do so much work, but that's a different topic (only made worse by adding more matrices jobs). At some point I think we should probably only make more intensive checks on pushing to the main branch or nightly.

daniel-zullo-frequenz committed:
Add job to check for all `test-installation` jobs

The added job runs if all the `test-installation` matrix jobs
ran and succeeded. The job is required to have a single job
that is used in branch protection rules so that updating the
protection rules is not needed each time we add or remove a job
from the matrix.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz commented:
> Most of the changes in this draft need to be pushed in [frequenz-floss/frequenz-repo-config-python](https://github.com/frequenz-floss/frequenz-repo-config-python) instead, excepts for dependencies updates.

There is already a ticket for it https://github.com/frequenz-floss/frequenz-repo-config-python/issues/168

llucax commented on .github/containers/nox-cross-arch/arm64-ubuntu-20.04-python-3.12.Dockerfile:
```diff
@@ -0,0 +1,33 @@
+# License: MIT
```
Can't we parameterize this file instead using [docker build variables](https://docs.docker.com/build/building/variables/)?

daniel-zullo-frequenz replied:
I think it should be possible. I'll give it a try

daniel-zullo-frequenz replied:
It works fine, suggestion addressed

llucax commented on .github/containers/test-installation/python-3.12.Dockerfile:
```diff
@@ -0,0 +1,17 @@
+# License: MIT
```
And this one.

daniel-zullo-frequenz replied:
done

llucax commented on .github/workflows/ci.yaml:
```diff
@@ -256,11 +266,24 @@ jobs:
         with:
           context: .
-          file: .github/containers/test-installation/Dockerfile
+          file: .github/containers/test-installation/python-${{ matrix.python }}.Dockerfile
```
In here you should probably pass an `arg:` to parameterize it.

daniel-zullo-frequenz replied:
I use ` build-args` instead. Done

llucax commented on .github/workflows/ci.yaml:
```diff
@@ -256,11 +266,24 @@ jobs:
           tags: localhost/test-installation
           push: false
 
+  # This job runs if all the `test-installation` matrix jobs ran and succeeded.
+  # It is only used to have a single job that we can require in branch
+  # protection rules, so we don't have to update the protection rules each time
+  # we add or remove a job from the matrix.
+  test-installation-all:
+    # The job name should match the name of the `test-installation` job.
+    name: Test package installation in different architectures
+    needs: ["test-installation"]
+    runs-on: ubuntu-20.04
+    steps:
+      - name: Return true
+        run: "true"
+
```
There is an updated fixed version of these "merging" jobs so merge queues can really check them. You can have a look at repo-config.

daniel-zullo-frequenz replied:
oh yes, I now remember about this as I reviewed it :/
I'll apply those changes here and probable other similar jobs if needed

daniel-zullo-frequenz replied:
Addressed

daniel-zullo-frequenz committed:
Update conditions for CI *-all jobs based on results

The current condition for the `*-all` jobs in the CI were
not correct, as it was skipping the job if the matrix job
succeeded, which is not the desired behavior. We want to
skip the *-all job only if the matrix job was skipped.
The results for `*-all` jobs are now set based on the
matrix job results.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
CI: Parameterize python version in test-installation

The python version used in the test-installation Dockerfile
was hardcoded and we ended up with a couple of similar
Dockerfiles for different python versions.
Ideally we should keep a single Dockerfile to use
different python versions.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
CI: Parameterize Python version in cross-arch Dockerfile

The python version in the Dockerfile for the arm64 cross-arch
tests is now parameterized to pass the python version as an
argument so that only one Dockerfile is needed for all python
versions.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz commented on .github/containers/nox-cross-arch/arm64-ubuntu-20.04-python.Dockerfile:
```diff
@@ -3,11 +3,13 @@
 # This Dockerfile is used to run the tests in arm64, which is not supported by
 # GitHub Actions at the moment.
 
+ARG PYTHON_VERSION="3.11"
```
@llucax I made a mistake here defining this arg in the global scope and not consuming it within the FROM stage. I think I can define it directly within the local scope of the stage

daniel-zullo-frequenz replied:
Fixed, but it will need a new approval

daniel-zullo-frequenz replied:
I tested the arm64 docker file in my local environment with different python versiona and it works smoothly

llucax replied:
Ah, good catch. Will approve.

daniel-zullo-frequenz commented:
Auto-merge is enabled. So please do not approve it if changes are required
